Support undefined grads in vmap fallback

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46674 Use new_zeros in evenly_distribute_backward
* **#46671 Support undefined grads in vmap fallback**

Previously, the vmap fallback would choke whenever it saw an undefined
tensor. For each sample in a batch, the fallback runs an operator
and then stacks together outputs to get the actual output.
Undefined tensors can occur as outputs while computing batched gradients
with vmap.

This PR updates the vmap fallback to handle undefined tensors which can
appear in backward formulas:
- if for each sample in a batch the output was undefined, then the vmap
fallback returns an undefined tensor
- if for each sample in a batch the output is defined, then the vmap
fallback stacks together the defined tensors
- if for some samples in a batch the output is defined/undefined, then
we error out.

Test Plan:
- new tests

Differential Revision: [D24454909](https://our.internmc.facebook.com/intern/diff/D24454909)

zou3519 committed:
Support undefined grads in vmap fallback

Previously, the vmap fallback would choke whenever it saw an undefined
tensor. For each sample in a batch, the fallback runs an operator
and then stacks together outputs to get the actual output.
Undefined tensors can occur as outputs while computing batched gradients
with vmap.

This PR updates the vmap fallback to handle undefined tensors which can
appear in backward formulas:
- if for each sample in a batch the output was undefined, then the vmap
fallback returns an undefined tensor
- if for each sample in a batch the output is defined, then the vmap
fallback stacks together the defined tensors
- if for some samples in a batch the output is defined/undefined, then
we error out.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Support undefined grads in vmap fallback"


Previously, the vmap fallback would choke whenever it saw an undefined
tensor. For each sample in a batch, the fallback runs an operator
and then stacks together outputs to get the actual output.
Undefined tensors can occur as outputs while computing batched gradients
with vmap.

This PR updates the vmap fallback to handle undefined tensors which can
appear in backward formulas:
- if for each sample in a batch the output was undefined, then the vmap
fallback returns an undefined tensor
- if for each sample in a batch the output is defined, then the vmap
fallback stacks together the defined tensors
- if for some samples in a batch the output is defined/undefined, then
we error out.

Test Plan:
- new tests

Differential Revision: [D24454909](https://our.internmc.facebook.com/intern/diff/D24454909)

[ghstack-poisoned]

albanD commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -195,6 +195,33 @@ void batchedTensorInplaceForLoopFallback(const c10::OperatorHandle& op, torch::j
+  TORCH_CHECK(false,
+      "vmap: slow fallback received a mix of undefined and defined tensors ",
+      "as the result of an operation. This is not supported. If this error ",
+      "message appears in framework code (e.g. autograd), please file us ",
```
nit: maybe we could encourage the user to always report an issue on github? I am very curious what they would do to hit that :p

zou3519 replied:
One thing a user could do is have a forward function that they created via c++ extensions that sometimes returns undefined tensors... but yeah, would be curious about the use case for anything like that

albanD commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -318,7 +345,12 @@ void batchedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Sta
-    auto flat_output = at::stack(shards);
+    auto flat_output = safeStack(shards);
+    // See NOTE [vmap through backward and undefined grad]
+    if (!flat_output.defined()) {
```
What's the deal with undefined Tensors and batch dimensions? Should they have some?

zou3519 replied:
BatchedTensor cannot hold an undefined tensor.

albanD replied:
Sounds good!

albanD commented on test/test_vmap.py:
```diff
@@ -733,6 +733,25 @@ def test_nn_module(self):
+            grad_x, = torch.autograd.grad(result, x, v)
+            return grad_x
+
+        # Runs vmap(get_vjp)(v), which should not error out.
```
Note that a simple way to generate undefined gradients is with `out = torch._C._functions.UndefinedGrad()(inp)` which is going to ignore the gradient flowing for out and return and undefined gradient for inp.
Or with a custom autograd.Function that returns `None` for a given input (you might actually be able to use a custom Function to test your error message easily actually as `None` are propagated as undefined Tensors properly now).

zou3519 replied:
vmap doesn't see a custom autograd.Function as an operator, so the vmap fallback won't run on the autograd.Function (it'll run on torch operations inside the Function). A better way to test this would be to create a dummy operator that has a derivative formula that returns undefined tensors

albanD replied:
Ho right you are not at the autograd level but at operator. Sounds good.

zou3519 committed:
Update on "Support undefined grads in vmap fallback"


Previously, the vmap fallback would choke whenever it saw an undefined
tensor. For each sample in a batch, the fallback runs an operator
and then stacks together outputs to get the actual output.
Undefined tensors can occur as outputs while computing batched gradients
with vmap.

This PR updates the vmap fallback to handle undefined tensors which can
appear in backward formulas:
- if for each sample in a batch the output was undefined, then the vmap
fallback returns an undefined tensor
- if for each sample in a batch the output is defined, then the vmap
fallback stacks together the defined tensors
- if for some samples in a batch the output is defined/undefined, then
we error out.

Test Plan:
- new tests

Differential Revision: [D24454909](https://our.internmc.facebook.com/intern/diff/D24454909)

[ghstack-poisoned]
